File: day-19/main.py
from ortools.linear_solver import pywraplp
import re

# PART1_FILE_NAME = "sample.txt"
PART1_FILE_NAME = "input.txt"
# PART2_FILE_NAME = "sample.txt"
PART2_FILE_NAME = "input.txt"


def f(blueprint, T):
    solver = pywraplp.Solver.CreateSolver("SCIP")
    if not solver:
        raise Exception("System Error")
    infinity = solver.infinity()

    # robots built each min
    o_build = {t: solver.IntVar(0.0, infinity, f"o_{t}") for t in range(2, T)}
    c_build = {t: solver.IntVar(0.0, infinity, f"c_{t}") for t in range(2, T)}
    b_build = {t: solver.IntVar(0.0, infinity, f"b_{t}") for t in range(2, T)}
    g_build = {t: solver.IntVar(0.0, infinity, f"g_{t}") for t in range(2, T)}

    # min 1
    o, c, b, g = 1, 0, 0, 0
    o_r, c_r, b_r, g_r = 1, 0, 0, 0
    for t in range(2, T):
        # build one at a time
        solver.Add(o_build[t] + c_build[t] + b_build[t] + g_build[t] <= 1)

        # assets used
        o -= (
            o_build[t] * blueprint["o_o"]
            + c_build[t] * blueprint["c_o"]
            + b_build[t] * blueprint["b_o"]
            + g_build[t] * blueprint["g_o"]
        )
        solver.Add(o >= 0)
        c -= b_build[t] * blueprint["b_c"]
        solver.Add(c >= 0)
        b -= g_build[t] * blueprint["g_b"]
        solver.Add(b >= 0)

        # proceed
        o, c, b, g = o + o_r, c + c_r, b + b_r, g + g_r
        o_r += o_build[t]
        c_r += c_build[t]
        b_r += b_build[t]
        g_r += g_build[t]
    g += g_r  # last minute

    solver.Maximize(g)
    status = solver.Solve()
    if status == pywraplp.Solver.OPTIMAL:
        return int(solver.Objective().Value() + 1e-6)
    raise Exception("No solution found")


def parse_line(line):
    x = re.search(
        r"Blueprint \d+: "
        + r"Each ore robot costs (\d+) ore. "
        + r"Each clay robot costs (\d+) ore. "
        + r"Each obsidian robot costs (\d+) ore and (\d+) clay. "
        + r"Each geode robot costs (\d+) ore and (\d+) obsidian.",
        line,
    )
    return {
        "o_o": int(x.group(1)),
        "c_o": int(x.group(2)),
        "b_o": int(x.group(3)),
        "b_c": int(x.group(4)),
        "g_o": int(x.group(5)),
        "g_b": int(x.group(6)),
    }


def main():
    ans = 0
    for idx, blueprint_str in enumerate(open(PART1_FILE_NAME, "r")):
        blueprint = parse_line(blueprint_str)
        res = f(blueprint, 24)
        # print(idx + 1, res)
        ans += (idx + 1) * res
    print("Part 1:", ans)

    ans = 1
    for idx, blueprint_str in enumerate(open(PART2_FILE_NAME, "r")):
        if idx >= 3:
            break
        blueprint = parse_line(blueprint_str)
        res = f(blueprint, 32)
        # print(idx + 1, res)
        ans *= res
    print("Part 2:", ans)


if __name__ == "__main__":
    main()

File: day-19/src/main.rs
use lazy_static::lazy_static;
use regex::Regex;
use std::fs;
use std::time::Instant;

// const PART1_FILENAME: &str = "sample.txt";
const PART1_FILENAME: &str = "input.txt";
const PART1_TIME: usize = 24;

// const PART2_FILENAME: &str = "sample.txt";
const PART2_FILENAME: &str = "input.txt";
const PART2_TIME: usize = 32;

#[derive(Debug, Clone, Copy)]
struct ASSETS {
    ore: usize,
    clay: usize,
    obsdn: usize,
    geode: usize,
}

#[derive(Debug, Clone, Copy)]
struct ROBOTS {
    ore: usize,
    clay: usize,
    obsdn: usize,
    geode: usize,
}

struct BLUEPRINT {
    ore_ore: usize,
    clay_ore: usize,
    obsdn_ore: usize,
    obsdn_clay: usize,
    geode_ore: usize,
    geode_obsdn: usize,
}

fn read_input(filename: &str) -> String {
    fs::read_to_string(filename).expect("Input for the problem")
}

fn cap_to_usize(caps: &regex::Captures, name: &str) -> usize {
    caps.name(name).unwrap().as_str().parse::<usize>().unwrap()
}

fn parse_line(line: &str) -> BLUEPRINT {
    lazy_static! {
        static ref RE: Regex = Regex::new(
            r"^Blueprint \d+: Each ore robot costs (?P<v1>\d+) ore. Each clay robot costs (?P<v2>\d+) ore. Each obsidian robot costs (?P<v3>\d+) ore and (?P<v4>\d+) clay. Each geode robot costs (?P<v5>\d+) ore and (?P<v6>\d+) obsidian.$"
        )
        .unwrap();
    }
    let caps = RE.captures(line).unwrap();
    BLUEPRINT {
        ore_ore: cap_to_usize(&caps, "v1"),
        clay_ore: cap_to_usize(&caps, "v2"),
        obsdn_ore: cap_to_usize(&caps, "v3"),
        obsdn_clay: cap_to_usize(&caps, "v4"),
        geode_ore: cap_to_usize(&caps, "v5"),
        geode_obsdn: cap_to_usize(&caps, "v6"),
    }
}

fn time_to_build(asset_have: usize, asset_need: usize, robot: usize) -> usize {
    if asset_have >= asset_need {
        0
    } else {
        ((asset_need - asset_have) + robot - 1) / robot
    }
}

fn branch_n_bound(
    blueprint: &BLUEPRINT,
    max_time: usize,
    t: usize,
    assets: &mut ASSETS,
    robots: &mut ROBOTS,
    res: &mut usize,
) -> bool {
    let mut o_assets = assets.clone();
    let mut o_robots = robots.clone();
    for _ in t + 1..max_time {
        let buildable_robots = ROBOTS {
            ore: o_assets.ore / blueprint.ore_ore,
            clay: o_assets.ore / blueprint.clay_ore,
            obsdn: std::cmp::min(
                o_assets.ore / blueprint.obsdn_ore,
                o_assets.clay / blueprint.obsdn_clay,
            ),
            geode: std::cmp::min(
                o_assets.ore / blueprint.geode_ore,
                o_assets.obsdn / blueprint.geode_obsdn,
            ),
        };
        o_assets.ore += o_robots.ore;
        o_assets.clay += o_robots.clay;
        o_assets.obsdn += o_robots.obsdn;
        o_assets.geode += o_robots.geode;
        if o_assets.geode > *res {
            return false;
        }
        o_robots.ore += buildable_robots.ore;
        o_robots.clay += buildable_robots.clay;
        o_robots.obsdn += buildable_robots.obsdn;
        o_robots.geode += buildable_robots.geode;
    }
    o_assets.geode += o_robots.geode;
    o_assets.geode <= *res
}

fn search(
    blueprint: &BLUEPRINT,
    max_time: usize,
    t: usize,
    assets: &mut ASSETS,
    robots: &mut ROBOTS,
    res: &mut usize,
) {
    if branch_n_bound(blueprint, max_time, t, assets, robots, res) {
        return;
    }
    // do nothing
    {
        let sub_res = (max_time - t) * robots.geode + assets.geode;
        if sub_res > *res {
            *res = sub_res;
        }
    }
    // next build
    {
        // geode robot
        if robots.obsdn > 0 {
            let dt = std::cmp::max(
                time_to_build(assets.ore, blueprint.geode_ore, robots.ore),
                time_to_build(assets.obsdn, blueprint.geode_obsdn, robots.obsdn),
            );
            if t + dt + 1 < max_time {
                assets.ore += (dt + 1) * robots.ore;
                assets.clay += (dt + 1) * robots.clay;
                assets.obsdn += (dt + 1) * robots.obsdn;
                assets.geode += (dt + 1) * robots.geode;

                assets.ore -= blueprint.geode_ore;
                assets.obsdn -= blueprint.geode_obsdn;
                robots.geode += 1;
                search(blueprint, max_time, t + dt + 1, assets, robots, res);
                assets.ore += blueprint.geode_ore;
                assets.obsdn += blueprint.geode_obsdn;
                robots.geode -= 1;

                assets.ore -= (dt + 1) * robots.ore;
                assets.clay -= (dt + 1) * robots.clay;
                assets.obsdn -= (dt + 1) * robots.obsdn;
                assets.geode -= (dt + 1) * robots.geode;
            }
        }
        // obsidian robot
        if robots.clay > 0 {
            let dt = std::cmp::max(
                time_to_build(assets.ore, blueprint.obsdn_ore, robots.ore),
                time_to_build(assets.clay, blueprint.obsdn_clay, robots.clay),
            );
            if t + dt + 1 < max_time {
                assets.ore += (dt + 1) * robots.ore;
                assets.clay += (dt + 1) * robots.clay;
                assets.obsdn += (dt + 1) * robots.obsdn;
                assets.geode += (dt + 1) * robots.geode;

                assets.ore -= blueprint.obsdn_ore;
                assets.clay -= blueprint.obsdn_clay;
                robots.obsdn += 1;
                search(blueprint, max_time, t + dt + 1, assets, robots, res);
                assets.ore += blueprint.obsdn_ore;
                assets.clay += blueprint.obsdn_clay;
                robots.obsdn -= 1;

                assets.ore -= (dt + 1) * robots.ore;
                assets.clay -= (dt + 1) * robots.clay;
                assets.obsdn -= (dt + 1) * robots.obsdn;
                assets.geode -= (dt + 1) * robots.geode;
            }
        }
        // clay robot
        let dt = time_to_build(assets.ore, blueprint.clay_ore, robots.ore);
        if t + dt + 1 < max_time {
            assets.ore += (dt + 1) * robots.ore;
            assets.clay += (dt + 1) * robots.clay;
            assets.obsdn += (dt + 1) * robots.obsdn;
            assets.geode += (dt + 1) * robots.geode;

            assets.ore -= blueprint.clay_ore;
            robots.clay += 1;
            search(blueprint, max_time, t + dt + 1, assets, robots, res);
            assets.ore += blueprint.clay_ore;
            robots.clay -= 1;

            assets.ore -= (dt + 1) * robots.ore;
            assets.clay -= (dt + 1) * robots.clay;
            assets.obsdn -= (dt + 1) * robots.obsdn;
            assets.geode -= (dt + 1) * robots.geode;
        }
        // ore robot
        let dt = time_to_build(assets.ore, blueprint.ore_ore, robots.ore);
        if t + dt + 1 < max_time {
            assets.ore += (dt + 1) * robots.ore;
            assets.clay += (dt + 1) * robots.clay;
            assets.obsdn += (dt + 1) * robots.obsdn;
            assets.geode += (dt + 1) * robots.geode;

            assets.ore -= blueprint.ore_ore;
            robots.ore += 1;
            search(blueprint, max_time, t + dt + 1, assets, robots, res);
            assets.ore += blueprint.ore_ore;
            robots.ore -= 1;

            assets.ore -= (dt + 1) * robots.ore;
            assets.clay -= (dt + 1) * robots.clay;
            assets.obsdn -= (dt + 1) * robots.obsdn;
            assets.geode -= (dt + 1) * robots.geode;
        }
    }
}

fn f(blueprint: &BLUEPRINT, max_time: usize) -> usize {
    let mut assets = ASSETS {
        ore: 1,
        clay: 0,
        obsdn: 0,
        geode: 0,
    };
    let mut robots = ROBOTS {
        ore: 1,
        clay: 0,
        obsdn: 0,
        geode: 0,
    };
    let mut res = 0usize;
    search(blueprint, max_time, 1, &mut assets, &mut robots, &mut res);
    res
}

fn part1() {
    let input = read_input(PART1_FILENAME);
    let mut res = 0;
    for (idx, blueprint) in input.split("\n").map(|line| parse_line(line)).enumerate() {
        let sub_res = f(&blueprint, PART1_TIME);
        // println!("{} {}", idx + 1, sub_res);
        res += (idx + 1) * sub_res;
    }
    println!("{}", res);
}

fn part2() {
    let input = read_input(PART2_FILENAME);
    let mut res = 1;
    for (idx, blueprint) in input.split("\n").map(|line| parse_line(line)).enumerate() {
        if idx >= 3 {
            break;
        }
        let sub_res = f(&blueprint, PART2_TIME);
        // println!("{} {}", idx + 1, sub_res);
        res *= sub_res;
    }
    println!("{}", res);
}

fn main() {
    let now = Instant::now();
    part1();
    println!("Part 1 elapsed: {:.2?}", now.elapsed());

    let now = Instant::now();
    part2();
    println!("Part 2 elapsed: {:.2?}", now.elapsed());
}
